From 367a9234d8de6bce173799d1e23fa9ec4664ee4b Mon Sep 17 00:00:00 2001
From: Noam Lessner <nlessner@habana.ai>
Date: Wed, 22 Apr 2020 13:50:50 +0300
Subject: [PATCH] [SW-11666] goya2: seperate ssc  sris srns straps

So we can select spread spectrum clocking (SSC), separate reference
clock with independent spread (SRIS) and separate reference clock
with no SSC (SRNS) we have 3 seperate boot straps that contoll this.

Change-Id: I2295d579f1a19745cdaba6e1b674005c34eb37ce
Signed-off-by: Noam Lessner <nlessner@habana.ai>
---

diff --git a/common/pcie.c b/common/pcie.c
index c7bbaaf..a3c18e7 100644
--- a/common/pcie.c
+++ b/common/pcie.c
@@ -29,25 +29,44 @@
 		regmap_writel(value, addr);
 }
 
-static void pcie_wa_snps_case_01038891(uint8_t ssc_en)
+static void pcie_wa_snps_case_01038891(uint32_t boot_strap)
 {
 	uint32_t reg = 0;
+	uint8_t separate_reference = 0;
+	uint8_t ssc_en = 0;
+
+	/* SRIS mode enable */
+	if (boot_strap &
+	    PSOC_GLOBAL_CONF_BOOT_STRAP_PINS_L_SRIS_MODE_MASK)
+		separate_reference = 1;
+
+	/* SRNS mode enable */
+	if ((boot_strap &
+	     PSOC_GLOBAL_CONF_BOOT_STRAP_PINS_L_BTL_EN_MASK) == 0)
+		separate_reference = 1;
+
+	/* spread spectum clock enable */
+	if (boot_strap &
+	    PSOC_GLOBAL_CONF_BOOT_STRAP_PINS_L_BTL_ROM_EN_MASK)
+		ssc_en = 1;
 
 	if ((arcboot_rom_global.config_fuse &
 	     EFUSE_SYSCFG_WA_SNPS_CASE_01038891) == 0) {
 		if (ssc_en) {
-			pcie_update_reg_range(
-				mmPCIE_PHY_PHYX_MPLLB_SSC_EN_0,
-				mmPCIE_PHY_PHYX_MPLLB_SSC_EN_1, 0x1);
-			pcie_update_reg_range(
-				mmPCIE_PHY_PHYX_MPLLA_SSC_EN_0,
-				mmPCIE_PHY_PHYX_MPLLA_SSC_EN_1, 0x1);
+			pcie_update_reg_range(mmPCIE_PHY_PHYX_MPLLB_SSC_EN_0,
+					mmPCIE_PHY_PHYX_MPLLB_SSC_EN_1, 0x1);
+			pcie_update_reg_range(mmPCIE_PHY_PHYX_MPLLA_SSC_EN_0,
+					mmPCIE_PHY_PHYX_MPLLA_SSC_EN_1, 0x1);
+			regmap_writel(0x5000, mmPCIE_PHY_EXT_MPLLB_SSC_PEAK);
+			regmap_writel(0x19CE,
+				      mmPCIE_PHY_EXT_MPLLB_SSC_STEP_SIZE);
+			regmap_writel(0x6666, mmPCIE_PHY_EXT_MPLLA_SSC_PEAK);
+			regmap_writel(0x2108,
+				      mmPCIE_PHY_EXT_MPLLA_SSC_STEP_SIZE);
 		}
 
-		pcie_update_reg_range(
-			mmPCIE_PHY_EXT_TX_DCC_CTRL_RANGE_G1,
-			mmPCIE_PHY_EXT_TX_DCC_CTRL_RANGE_G4, 0xC);
-
+		pcie_update_reg_range(mmPCIE_PHY_EXT_TX_DCC_CTRL_RANGE_G1,
+				mmPCIE_PHY_EXT_TX_DCC_CTRL_RANGE_G4, 0xC);
 		pcie_update_reg_range(
 			mmPCIE_PHY_PROTOCOL0_EXT_RX_CDR_VCO_CONFIG_G1_0,
 			mmPCIE_PHY_PROTOCOL0_EXT_RX_CDR_VCO_CONFIG_G1_7, 0x1);
@@ -55,32 +74,20 @@
 			mmPCIE_PHY_PROTOCOL0_EXT_RX_CDR_VCO_CONFIG_G2_0,
 			mmPCIE_PHY_PROTOCOL0_EXT_RX_CDR_VCO_CONFIG_G2_7, 0x1);
 
-		if (ssc_en) {
+		if (separate_reference) {
 			reg = 0xA0;
-		} else {
-			reg = 0x50;
-		}
-		regmap_writel(reg, mmPCIE_PHY_PROTOCOL0_EXT_MPLLA_MULTIPLIER);
-		regmap_writel(0x7D, mmPCIE_PHY_PROTOCOL0_EXT_MPLLB_MULTIPLIER);
-
-		if (ssc_en) {
 			regmap_writel(0x00250025,
 				      mmPCIE_PHY_PROTOCOL0_EXT_MPLLA_BANDWIDTH);
 			regmap_writel(0x00250025,
 				      mmPCIE_PHY_PROTOCOL0_EXT_MPLLB_BANDWIDTH);
-			regmap_writel(0x5000,
-				      mmPCIE_PHY_EXT_MPLLB_SSC_PEAK);
-			regmap_writel(0x19CE,
-				      mmPCIE_PHY_EXT_MPLLB_SSC_STEP_SIZE);
-			regmap_writel(0x6666,
-				      mmPCIE_PHY_EXT_MPLLA_SSC_PEAK);
-			regmap_writel(0x2108,
-				      mmPCIE_PHY_EXT_MPLLA_SSC_STEP_SIZE);
 		} else {
+			reg = 0x50;
 			regmap_writel(0x01C301C3,
 				      mmPCIE_PHY_PROTOCOL0_EXT_MPLLB_BANDWIDTH);
 		}
 
+		regmap_writel(reg, mmPCIE_PHY_PROTOCOL0_EXT_MPLLA_MULTIPLIER);
+		regmap_writel(0x7D, mmPCIE_PHY_PROTOCOL0_EXT_MPLLB_MULTIPLIER);
 		regmap_writel(0x30,
 			      mmPCIE_PHY_PROTOCOL0_EXT_TX_EQ_MAIN_G1_SLICE0);
 		regmap_writel(0x30,
@@ -100,7 +107,6 @@
 static void pcie_config(void)
 {
 	uint8_t sram_init_done;
-	uint8_t ssc_en = 0;
 	uint32_t boot_strap, flr_ctrl, link_ctrl;
 	uint8_t ltssm_en;
 
@@ -120,11 +126,7 @@
 		    PSOC_GLOBAL_CONF_BOOT_STRAP_PINS_L_SRIS_MODE_MASK)
 			regmap_writel(1, mmPCIE_PHY_PIPE_RX_SRIS_MODE_EN);
 
-		if (boot_strap &
-		    PSOC_GLOBAL_CONF_BOOT_STRAP_PINS_L_BTL_EN_MASK)
-			ssc_en= 1;
-
-		pcie_wa_snps_case_01038891(ssc_en);
+		pcie_wa_snps_case_01038891(boot_strap);
 		regmap_writel(1, mmPSOC_GLOBAL_CONF_PCIE_PHY_RST_N);
 
 		sram_init_done = 0;
